File: src/main/java/com/jcloisterzone/game/capability/MarketplaceCapability.java
package com.jcloisterzone.game.capability;

import com.jcloisterzone.Player;
import com.jcloisterzone.board.Position;
import com.jcloisterzone.board.pointer.FeaturePointer;
import com.jcloisterzone.board.pointer.ScorePositionsFeaturePointer;
import com.jcloisterzone.event.ExprItem;
import com.jcloisterzone.event.PointsExpression;
import com.jcloisterzone.event.ScoreEvent;
import com.jcloisterzone.event.ScoreEvent.ReceivedPoints;
import com.jcloisterzone.feature.Completable;
import com.jcloisterzone.feature.Marketplace;
import com.jcloisterzone.feature.Scoreable;
import com.jcloisterzone.feature.Road;
import com.jcloisterzone.figure.Follower;
import com.jcloisterzone.game.Capability;
import com.jcloisterzone.game.state.GameState;

import io.vavr.collection.HashSet;
import io.vavr.collection.List;
import io.vavr.collection.Set;
import io.vavr.Tuple2;

import java.util.ArrayList;

public class MarketplaceCapability extends Capability<Void> {

    @Override
    public List<ReceivedPoints> appendFiguresBonusPoints(GameState state, List<ReceivedPoints> bonusPoints, Scoreable feature, boolean isFinal) {
        if (feature instanceof Road && !isFinal) {
        	Road road = (Road) feature;
			Set<Player> owners = road.getOwners(state);
    		Set<FeaturePointer> fps = road.getMarketplaces();
    		if (fps.size()>0 && owners.size()>0) {
    			Integer tiles = 0;
    			Set<Position> positions = HashSet.empty();
    			for(FeaturePointer fp : fps) {
    				Marketplace marketplace = (Marketplace) state.getFeatureMap()
    					.get(fp.getPosition())
        			    .flatMap(m -> m.get(fp))
        			    .get();
            		List<Road> marketplaceRoads = marketplace.getMarketplaceRoads(state);
            		for(Road marketplaceRoad: marketplaceRoads) {
                		Set<Position> roadPositions = marketplaceRoad.getTilePositions();
                		positions = positions.addAll(roadPositions);
//    					if (!road.equals(marketplaceRoads)) {
    						tiles+= roadPositions.size();
//    					}
    				}
    				tiles-= road.getTilePositions().size(); // Exclude scored road tiles
    			}
                ArrayList<ExprItem> exprItems = new ArrayList<ExprItem>();
                ExprItem expr = new ExprItem(tiles, "tiles", tiles);
                for (Player player: owners) {
                    List<Tuple2<Follower, FeaturePointer>> followers = feature.getFollowers2(state).filter(t -> t._1.getPlayer().equals(player)).toList();
                    bonusPoints = bonusPoints.append(new ScoreEvent.ReceivedPoints(new PointsExpression("marketplace", expr), player, new ScorePositionsFeaturePointer(followers.get()._2, positions))); // Positions or Feature pointer
                }
        	}
        }
        return bonusPoints;
    }

}

File: src/main/ts/com/jcloisterzone/game/capability/MarketplaceCapability.ts
import { HashSet, type Set } from "../../../../io/vavr/Set.js";
import type { Position } from "../../board/Position.js";
import type { FeaturePointer } from "../../board/pointer/FeaturePointer.js";
import type { Marketplace } from "../../feature/Marketplace.js";
import type { Road } from "../../feature/Road.js";
import { Capability } from "../Capability.js";
import type { GameState } from "../state/GameState.js";

/** Marketplace (Bridges, Castles &amp; Bazaars) scoring helper. */
export class MarketplaceCapability extends Capability<void> {
  // NOTE: appendFiguresBonusPoints is commented out in the Java source too.

  getMarketplaceOtherRoadsTiles(
    state: GameState,
    road: Road,
    marketplaceFp: FeaturePointer,
    completed: boolean,
  ): number | null {
    if (completed) {
      let tiles = 0;
      let positions: Set<Position> = HashSet.empty<Position>();
      const marketplace = state.getFeature(marketplaceFp) as Marketplace;
      const marketplaceRoads = marketplace.getMarketplaceRoads(state);
      for (const marketplaceRoad of marketplaceRoads) {
        const roadPositions = marketplaceRoad.getTilePositions();
        positions = positions.addAll(roadPositions);
        tiles += roadPositions.size();
      }
      tiles -= road.getTilePositions().size(); // exclude scored road tiles
      return tiles;
    }
    return null;
  }
}

Capability.register(MarketplaceCapability);
